Add Joyent Triton discovery

richardkiene opened the pull request:
This commit adds support for discovering containers running on a Joyent Triton cloud (private or public).

I have implemented a Prometheus compatible exporter for Triton. The architecture is described in [RFD 27](https://github.com/joyent/rfd/blob/master/rfd/0027/README.md) and [RFD 37](https://github.com/joyent/rfd/blob/master/rfd/0037/README.md). You can also take a look at the source by visiting [triton-cmon](https://github.com/joyent/triton-cmon) and [triton-cmon-agent](https://github.com/joyent/triton-cmon-agent)

If you'd like to verify the functionality of this commit, you may do so by [running Triton in a VM](https://github.com/joyent/triton/blob/master/docs/developer-guide/coal-setup.md) and then following the [setup guide](https://github.com/joyent/triton-cmon/blob/CMON-8/docs/experimental_coal_install.md).

Alternatively, you can reach out to me and I will get you access to our east-3b public cloud beta deployment which has everything running already. The east-3b setup guide is located [here](https://gist.github.com/richardkiene/3b9d8537e0c4ac4df0eb75ca1fbff32e)

beorn7 commented:
Thanks for your contribution.

Note that we have to apply a pretty high bar for adding more SD mechanisms to the Prometheus codebase. We strongly prefer supporting more SD mechanisms via the [file-based SD](https://prometheus.io/docs/operating/configuration/#<file_sd_config>). Have a look at #2221 for a similar discussion.

Essentially we need to be convinced that…
* …you will be available to maintain the code and not just disappear at some point (as many others did, even after having pledged to maintain the code.)
* …the SD mechanism in question is in common use and mature enough to justify direct integration.
* …there is a technical advantage of direct integration (like the Kubernetes watches), i.e. the code in the main Prometheus binary is more than essentially a cron job moved into the Prometheus binary.

Please consider the points above and the discussion in PR #2221 and then follow up here with your conclusions.

The person most involved with SD and thus the most qualified gate keeper is probably @fabxc.

richardkiene commented:
> ...you will be available to maintain the code and not just disappear at some point (as many others did, even after having pledged to maintain the code.)

Understood, and we ([Joyent](https://www.joyent.com/) a fellow [CNCF member](https://www.joyent.com/blog/the-foundation-of-cloud-native-computing)) have no intention of abandoning this code, quite the opposite actually. Triton Service Discovery is an integral piece of our [monitoring and metrics offering](https://github.com/joyent/rfd/blob/master/rfd/0027/README.md).

With that said, I believe the most compelling case I can make for non-abandonment is that we're integrating this into our mature and widely used [product](https://www.joyent.com/triton), which is also an open source [project](https://github.com/joyent/triton). Additionally, as CNCF members ourselves, we have a compelling reason to support and help maintain Prometheus.

> ...the SD mechanism in question is in common use and mature enough to justify direct integration.

Myself, colleagues, and a few early adopting customers have been using this SD mechanism for a while now. Unfortunately, without promoting our fork (which seems counterproductive since our intention is to help promote and foster usage of Prometheus), I'm not sure how it is possible to satisfy this requirement without actually being included in Prometheus already.

> …there is a technical advantage of direct integration (like the Kubernetes watches), i.e. the code in the main Prometheus binary is more than essentially a cron job moved into the Prometheus binary.

The technical advantage of direct integration is that authentication and authorization are greatly simplified, as is labeling support. This is, however, on a timer because it was modeled after the SD mechanisms that had already been included in Prometheus many months ago. I could expose triggered events to Prometheus (similar to the Kubernetes watches functionality that is described in PR #2221) and remove the timer in favor of a push mechanism if that makes integration more compelling.

In PR #2221, they mention a lack of integration tests as a blocker. I would be happy to add integration tests to this PR. I actually have my own set of integration tests for our discovery endpoint and the main reason for not including integration tests with this PR is that I saw no clear example of how you and the other maintainers would like that to be done.

Thank you for your help and consideration. How shall we proceed?

juliusv commented:
>> ...the SD mechanism in question is in common use and mature enough to justify direct integration.
>
> Myself, colleagues, and a few early adopting customers have been using this SD mechanism for a while now. Unfortunately, without promoting our fork (which seems counterproductive since our intention is to help promote and foster usage of Prometheus), I'm not sure how it is possible to satisfy this requirement without actually being included in Prometheus already.

To clarify, @beorn7 meant general use of Triton's SD, not Prometheus-specific. So: are there many people out there in the world who use Triton's SD mechanism to discover services already, independently of Prometheus?

richardkiene commented:
@juliusv Yes, there are a large number of people who use our endpoints for service discovery. Either via the cert based endpoint that this PR consumes or our HTTP-signature based endpoint.

Examples:

- [Hashicorp](https://www.hashicorp.com/) [Terraform Triton Provider](https://www.terraform.io/docs/providers/triton/)
-[Autopilot](http://autopilotpattern.io/)
-[Docker](https://docker.com) via [sdc-docker](https://github.com/joyent/sdc-docker)

We also have a number of very large reference users (some of them customers, some of them open source) and I can investigate which ones I can name publicly if that is needed.

juliusv commented:
Thanks, those points (including willingness to maintain the code and supplying integration tests) all point in a good direction already. The decision here still lies with @fabxc and @beorn7 though.

fabxc commented on config/config.go:
```diff
@@ -1086,6 +1096,47 @@ func (c *AzureSDConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
+type TritonSDConfig struct {
+	Account            string         `yaml:"account"`
+	Cert               string         `yaml:"cert"`
+	DnsSuffix          string         `yaml:"dns_suffix"`
```
We keep uppercases for abbreviations: `DNSSuffix`

fabxc commented on config/config.go:
```diff
@@ -1086,6 +1096,47 @@ func (c *AzureSDConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	Cert               string         `yaml:"cert"`
+	DnsSuffix          string         `yaml:"dns_suffix"`
+	Endpoint           string         `yaml:"endpoint"`
+	InsecureSkipVerify bool           `yaml:"insecure_skip_verify"`
```
Anything speaking against using our standard TLS config object? https://github.com/prometheus/prometheus/blob/master/config/config.go#L963

We also have a helper to generate an HTTP client for it already: https://github.com/prometheus/prometheus/blob/master/util/httputil/client.go#L122-L152

richardkiene replied:
No, in fact, I would much rather share that configuration. I wasn't sure if that would be acceptable or not so I went ahead with my own. I'll happily remove mine and use the TLS config object. Thanks for pointing this out!

fabxc commented on discovery/triton/triton.go:
```diff
@@ -0,0 +1,150 @@
+
+// TritonDiscovery periodically performs Triton-SD requests. It implements
+// the TargetProvider interface.
+type TritonDiscovery struct {
```
As this is a dedicated package already, let's simplify this to `Discovery` so it's simply `triton.Discovery` for importers. Similar for `TritionDiscoveryResponse` etc.

fabxc commented on discovery/triton/triton.go:
```diff
@@ -0,0 +1,150 @@
+		return tg, fmt.Errorf("Error reading response body: %s", err)
+	}
+	tdr := TritonDiscoveryResponse{}
+	json.Unmarshal([]byte(string(data)), &tdr)
```
Unnecessary conversion from `[]byte` to `string` and back. The returned error should be handled.

Probably easiest to just use `json.Decoder` and skip the `ReadAll`.

fabxc commented on retrieval/discovery/triton.go:
```diff
@@ -0,0 +1,150 @@
+// Copyright 2016 The Prometheus Authors
```
Duplicate file for the location before we moved `discovery` into a top-level package?

richardkiene replied:
Darn! Yes, that is what happened. I'll remove this.

fabxc commented on discovery/triton/triton.go:
```diff
@@ -0,0 +1,150 @@
+	"net/http"
+	"time"
+
+	promlog "github.com/prometheus/common/log"
```
We just import this as `log` throughout our code base. See below regarding use of stdblib `log`.

fabxc replied:
So far we are using the global logger mostly. Ideally we want to have structured logging with contextualised loggers being passed down. The Kubernetes SD follows that pattern it would be great if any new SDs would as well: https://github.com/prometheus/prometheus/blob/master/discovery/kubernetes/kubernetes.go#L79

fabxc commented on discovery/triton/triton.go:
```diff
@@ -0,0 +1,150 @@
+func NewDiscovery(conf *config.TritonSDConfig) *TritonDiscovery {
+	cert, err := tls.LoadX509KeyPair(conf.Cert, conf.Key)
+	if err != nil {
+		log.Fatalln("Unable to load cert", err)
```
Exiting just because a single SD failed to start is not the greatest solution.

We have other SDs that can error as well, e.g.:https://github.com/prometheus/prometheus/blob/master/discovery/discovery.go#L98-L105

It's not perfect, but just logging in a single place and continuing with best effort is the better option.

richardkiene replied:
Fair enough, I'll change that.

richardkiene commented:
Hey @fabxc and @beorn7, sorry that took so long. I believe I've addressed all of your review and I've added tests for Triton Discovery as well as added Triton Discovery to the config tests.

Let me know if I missed something. Thanks in advance for the review!

fabxc commented on config/config.go:
```diff
@@ -1086,6 +1095,39 @@ func (c *AzureSDConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	DNSSuffix       string         `yaml:"dns_suffix"`
+	Endpoint        string         `yaml:"endpoint"`
+	Port            int            `yaml:"port"`
+	RefreshInterval model.Duration `yaml:"refresh_interval,omitempty"`
```
Has to be validated to not be zero. The instantiation of the ticker further down will panic otherwise.

richardkiene replied:
Check added on line 1128

fabxc commented on discovery/discovery.go:
```diff
@@ -106,6 +107,14 @@ func ProvidersFromConfig(cfg config.ServiceDiscoveryConfig) map[string]TargetPro
 		app("azure", i, azure.NewDiscovery(c))
 	}
+	for i, c := range cfg.TritonSDConfigs {
+		t, err := triton.New(log.Base(), c)
```
I know it's equally wrong in the other SDs, but let's make it `log.With("sd", "triton")` here so all log messages have proper context.

fabxc commented on discovery/triton/triton.go:
```diff
@@ -0,0 +1,169 @@
+	// Get an initial set right away.
+	tg, err := d.refresh()
+	if err != nil {
+		d.logger.Error(err)
```
Similarly, it should ideally be `d.logger.With("err", err).Error("Refreshing targets failed")`. Same a few lines down.

fabxc commented on discovery/triton/triton.go:
```diff
@@ -0,0 +1,169 @@
+
+	resp, err := d.client.Get(endpoint)
+	if err != nil {
+		return tg, fmt.Errorf("Error reaching endpoint: %s", err)
```
We keep error messages lower case and only do proper sentences in logging messages.
Not sure it's perfect, but it would be consistent.

fabxc commented:
Thanks for the fixes. LGTM after comments.

fabxc commented:
👍 thanks.

To complete this, could you file a PR against the next-release branch of prometheus/docs adding the SD to https://github.com/prometheus/docs/blob/master/content/docs/operating/configuration.md?

The SD should be marked as "beta state" like several others.

richardkiene commented:
@fabxc yep, can do. Thanks!
